docs/settings.rst

```diff
@@ -1771,6 +1774,18 @@ IP Address for webserver/API to listen on.
 
 Webserver/API access is only allowed from these subnets.
 
+.. _setting-webserver-hash-plaintext-credentials:
+
+``webserver-hash-plaintext-credentials``
+----------------------------------------
+..versionadded:: 4.6.0
+
+-  Boolean
+-  Default: no
+
+Whether passwords and API keys supplied in the configuration as plaintext should be hashed during startup, to prevent the plaintext versions from staying in memory. Doing so increases significantly the cost of verifying credentials and is thus disabled by default.
```

Comment:
I'm concerned about this unsafe default; why is it even an option to keep plain-text passwords in memory? How "significant" is the cost in terms of absolute time and/or memory? Would it be more sensible to instead make the hashing algorithm and/or tunable parameters configurable for constrained hosts?

Reply:
This has been extensively discussed in earlier comments.

Reply:
I saw https://github.com/PowerDNS/pdns/pull/10157#issuecomment-816516664 and a couple of comments suggesting that a previous implementation imposed a slowdown of 110 to 170 milliseconds per API request, which hardly seem "extensive". Nor could I find any decision regarding my primary point here about the default behavior remaining insecure, although again there was discussion that seemed to peter out.

Reply:
That's actually very expensive, given that simply loading the main page of the recursor does dozens of such queries, refreshes itself every second doing on average 2 requests per second. So that means we would spent half a second of the time of a CPU core to refresh that page, which in my book is a huge issue in a software whose main goal is to serve DNS queries. Does that sound like a good idea to you?
Now, the default is to use hashed webserver passwords and API keys if the administrator uses that in the configuration file. If they don't, keeping the plaintext in memory is not ideal, sure, but there isn't a lot of actual thread models where it does matter. An attacker would need to find some remote infoleak flaw and then, in most cases, they would be able to access read-only metrics and a random secret that is not used anywhere else. If you are worried about that, great, there is an option to hash these credentials in memory, just enable it.

Reply:
I think you misunderstood me. I said that a few comments' worth of conversation does not qualify as "extensive", not that a couple hundred milliseconds per request isn't _expensive_ (although making dozens of requests per page load seems like a problem in itself). But to rephrase my last question from above, why are the only options plain text vs. "huge issue" hashing? There's **lots** of software used for much more sensitive access control that gets by with far less computationally expensive authorization, and even in the DNS space BIND `rdnc` suffers from many other problems but avoids this one altogether.

Reply:
Sure, how would you do it? I'd be happy to review your pull request anytime.

Reply:
Default to a suitably fast hashing algorithm and don't allow opting out, but support opting _in_ to use of intentionally expensive key-stretching alternatives such as scrypt. Since supporting SHA-256 is already a requirement for DNSSEC anyway, that seems like a good choice.

> I'd be happy to review your pull request anytime.

This pull request is almost there now; I'm just arguing for safer defaults. But is that approach even on the table? If not then there's no point in preparing it.

Reply:
> Default to a suitably fast hashing algorithm and don't allow opting out, but support opting in to use of intentionally expensive key-stretching alternatives such as scrypt. Since supporting SHA-256 is already a requirement for DNSSEC anyway, that seems like a good choice.

That sounds good. One issue is that SHA-256 support is not a requirement for dnsdist so we need to deal with the case where it's not available. It might be nice to use `crypto_generichash` from libsodium when available, then SHA-256 from libcrypto, but we still need a fallback when none of this is available.

> This pull request is almost there now; I'm just arguing for safer defaults. But is that approach even on the table? If not then there's no point in preparing it.

Yes, I would be happy to have what you suggest. It would require writing more code than this PR currently have, so I might merge that PR first as it already improves things significantly, but I would definitely merge a suitable follow-up PR.